kafka-client/src/test/java/dev/responsive/kafka/async/AsyncProcessorIntegrationTest.java

```diff
@@ -0,0 +1,293 @@
+import org.apache.kafka.streams.processor.api.FixedKeyProcessorSupplier;
+import org.apache.kafka.streams.processor.api.FixedKeyRecord;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.StoreBuilder;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.TestInfo;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+public class AsyncProcessorIntegrationTest {
+
+  @RegisterExtension
+  static ResponsiveExtension EXTENSION = new ResponsiveExtension(StorageBackend.CASSANDRA);
+
+  private static final String INPUT_TOPIC = "input";
+  private static final String OUTPUT_TOPIC = "output";
+  private static final String ASYNC_KV_STORE = "async-kv";
+
+  private final Map<String, Object> responsiveProps = new HashMap<>();
+
+  private String name;
+  private Admin admin;
+
+  @BeforeEach
+  public void before(
+      final TestInfo info,
+      final Admin admin,
+      @ResponsiveConfigParam final Map<String, Object> responsiveProps
+  ) {
+    // add displayName to name to account for parameterized tests
+    name = info.getDisplayName().replace("()", "");
+
+    this.responsiveProps.putAll(responsiveProps);
+
+    this.admin = admin;
+    createTopicsAndWait(admin, Map.of(inputTopic(), 10, outputTopic(), 1));
+  }
+
+  @AfterEach
+  public void after() {
+    admin.deleteTopics(List.of(inputTopic(), outputTopic()));
+  }
+
+  private String inputTopic() {
+    return name + "." + INPUT_TOPIC;
+  }
+
+  private String outputTopic() {
+    return name + "." + OUTPUT_TOPIC;
+  }
+
+  @Test
+  public void shouldProcessEventsInOrderByKey() throws Exception {
+    // Given:
+    final List<String> keys = List.of("a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k",
+                                      "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v",
+                                      "w", "x", "y", "z");
+
+    // produce 5 records for each key, with all keys intertwined between iterations
+    final List<KeyValue<String, String>> inputRecords = new LinkedList<>();
+    for (int val = 1; val < 6; ++val) {
+      for (final String key : keys) {
+        inputRecords.add(new KeyValue<>(key, key + val));
+      }
+    }
+
+    final Map<String, String> finalOutputRecords = new HashMap<>(keys.size());
+    for (final String key : keys) {
+      final String finalValue = String.format("%s1%s2%s3%s4%s5", key, key, key, key, key);
+      finalOutputRecords.put(key, finalValue);
+    }
+
+    final Map<String, Object> properties = getMutableProperties();
+    final KafkaProducer<String, String> producer = new KafkaProducer<>(properties);
+    
+    final AtomicInteger processed = new AtomicInteger(0);
+    final Map<String, String> latestValues = new ConcurrentHashMap<>();
+
+    final StreamsBuilder builder = new StreamsBuilder();
+
+    final KStream<String, String> input = builder.stream(inputTopic());
```

Comment:
what size thread pool will this app use? lets make sure it's got a few threads running.

Reply:
The latest iteration of the test has 5 stream threads, each of which has 5 async threads. And enough keys to have multiple keys per partition and multiple partitions per streamthread

I'll document all this somewhere tomorrow